cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -344,36 +306,82 @@ func performEtcdStaticPodUpgrade(client clientset.Interface, waiter apiclient.Wa
 
 	// Checking health state of etcd after the upgrade
 	fmt.Println("[upgrade/etcd] Waiting for etcd to become available")
-	if _, err = newEtcdClient.WaitForClusterAvailable(retries, retryInterval); err != nil {
+	if _, err = newEtcdClient.WaitForClusterAvailable(etcdConnectRetries, etcdConnectRetryInterval); err != nil {
 		fmt.Printf("[upgrade/etcd] Failed to healthcheck etcd: %v\n", err)
-		// Despite the fact that upgradeComponent was successful, there is something wrong with the etcd cluster
-		// First step is to restore back up of datastore
-		fmt.Println("[upgrade/etcd] Rolling back etcd data")
-		if err := rollbackEtcdData(cfg, pathMgr); err != nil {
-			// Even copying back datastore failed, no options for recovery left, bailing out
-			return true, errors.Wrapf(err, "fatal error rolling back local etcd cluster datadir, the backup of etcd database is stored here:(%s)", backupEtcdDir)
+		if err := rollbackEtcdOnFailedUpgrade(pathMgr, cfg, recoverManifests, oldEtcdClient, backupEtcdDir, err); err != nil {
+			return true, err
 		}
-		fmt.Println("[upgrade/etcd] Etcd data rollback successful")
 
+		// We've successfully rolled back etcd, and now return an error describing that the upgrade failed
+		return true, errors.Wrap(err, "fatal error upgrading local etcd cluster, rolled the state back to pre-upgrade state")
+	}
+
+	return false, nil
+}
+
+// compareEtcdVersion is used to compare the currentEtcdVersion and desiredEtcdVersion
+func compareEtcdVersion(cfg *kubeadmapi.InitConfiguration, oldEtcdClient etcdutil.ClusterInterrogator) (bool, error) {
+	// Need to check currently used version and version from constants, if differs then upgrade
+	desiredEtcdVersion, err := constants.EtcdSupportedVersion(cfg.KubernetesVersion)
+	if err != nil {
+		return true, errors.Wrap(err, "failed to retrieve an etcd version for the target Kubernetes version")
+	}
+
+	// gets the etcd version of the local/stacked etcd member running on the current machine
+	currentEtcdVersions, err := oldEtcdClient.GetClusterVersions()
+	if err != nil {
+		return true, errors.Wrap(err, "failed to retrieve the current etcd version from oldEtcdClient")
+	}
+	currentEtcdVersionStr, ok := currentEtcdVersions[etcdutil.GetClientURL(&cfg.LocalAPIEndpoint)]
+	if !ok {
+		err = errors.Errorf("unknown LocalAPIEndpoint: %v", cfg.LocalAPIEndpoint)
```

Comment:
No need to assign to `err` here. You can just `return true, errors.Wrap(errors.Errorf("unknown LocalAPIEndpoint: %v", cfg.LocalAPIEndpoint), "xxx")` directly.

Reply:
Okay, thank you for your advice.